python/semantic_kernel/planning/stepwise_planner/stepwise_planner.py

```diff
@@ -0,0 +1,459 @@
+
+
+def read_file(file_path: str) -> str:
+    with open(file_path, "r") as file:
+        return file.read()
+
+
+# TODO: Original C# uses "StepwisePlanner_Excluded" for RESTRICTED_SKILL_NAME
+RESTRICTED_SKILL_NAME = "StepwisePlanner"
+S_FINAL_ANSWER_REGEX = re.compile(r"\[FINAL ANSWER\](?P<final_answer>.+)", re.DOTALL)
+S_THOUGHT_REGEX = re.compile(
+    r"(\[THOUGHT\])?(?P<thought>.+?)(?=\[ACTION\]|$)", re.DOTALL
+)
+S_ACTION_REGEX = re.compile(r"\[ACTION\][^{}]*({(?:[^{}]*{[^{}]*})*[^{}]*})", re.DOTALL)
+
+ACTION = "[ACTION]"
+THOUGHT = "[THOUGHT]"
+OBSERVATION = "[OBSERVATION]"
+SCRATCH_PAD_PREFIX = (
+    "This was my previous work (but they haven't seen any of it!"
+    " They only see what I return as final answer):"
+)
+
+
+def is_null_or_empty(value: str) -> bool:
+    return value is None or value == ""
+
+
+class StepwisePlanner:
+    config: StepwisePlannerConfig
+    _context: "SKContext"
+    _function_flow_function: "SKFunctionBase"
+
+    def __init__(
+        self,
+        kernel: Kernel,
+        config: StepwisePlannerConfig = None,
+        prompt: str = None,
+        prompt_user_config: PromptTemplateConfig = None,
+    ):
+        assert isinstance(kernel, Kernel)
+        self._kernel = kernel
+
+        self.config = config or StepwisePlannerConfig()
+        self.config.excluded_skills.append(RESTRICTED_SKILL_NAME)
+
+        prompt_config = prompt_user_config or PromptTemplateConfig()
+        prompt_template = prompt or read_file(PROMPT_TEMPLATE_FILE_PATH)
+
+        if prompt_user_config is None:
+            prompt_config = PromptTemplateConfig.from_json(
+                read_file(PROMPT_CONFIG_FILE_PATH)
+            )
+
+        prompt_config.completion.max_tokens = self.config.max_tokens
+
+        self._system_step_function = self.import_semantic_function(
+            kernel, "StepwiseStep", prompt_template, prompt_config
+        )
+        self._native_functions = self._kernel.import_skill(self, RESTRICTED_SKILL_NAME)
+
+        self._context = kernel.create_new_context()
+        self._logger = self._kernel.logger
+
+    def create_plan(self, goal: str) -> Plan:
+        if is_null_or_empty(goal):
+            raise PlanningException(
+                PlanningException.ErrorCodes.InvalidGoal, "The goal specified is empty"
+            )
+
+        function_descriptions = self.get_function_descriptions()
+
+        plan_step: Plan = Plan.from_function(self._native_functions["ExecutePlan"])
+        plan_step.parameters.set("function_descriptions", function_descriptions)
+        plan_step.parameters.set("question", goal)
+
+        plan_step._outputs.append("agent_scratch_pad")
+        plan_step._outputs.append("step_count")
+        plan_step._outputs.append("skill_count")
+        plan_step._outputs.append("steps_taken")
+
+        plan = Plan(goal)
+
+        plan.add_steps([plan_step])
+
+        return plan
+
+    # TODO: sync C# with https://github.com/microsoft/semantic-kernel/pull/1195
```

Comment:
Will update this comment thread when future dotnet changes are available for review.

Reply:
#2504